proxy/envoy/egress.go

```diff
@@ -0,0 +1,185 @@
+	if err := ctl.AppendServiceHandler(func(*model.Service, model.Event) {
+		out.reload()
+	}); err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
+
+func (w *egressWatcher) reload() {
+	w.agent.ScheduleConfigUpdate(generateEgress(w.context))
+}
+
+func (w *egressWatcher) Run(stop <-chan struct{}) {
+	go w.agent.Run(stop)
+
+	// Initialize envoy according to the current model state,
+	// instead of waiting for the first event to arrive.
+	// Note that this is currently done synchronously (blocking),
+	// to avoid racing with controller events lurking around the corner.
+	// This can be improved once we switch to a mechanism where reloads
+	// are linearized (e.g., by a single goroutine reloader).
+	w.reload()
+	w.ctl.Run(stop)
+}
+
+// EgressConfig defines information for engress
+type EgressConfig struct {
+	Namespace string
+	Services  model.ServiceDiscovery
+	Mesh      *config.ProxyMeshConfig
+}
+
+func generateEgress(conf *EgressConfig) *Config {
+
+	// Create a VirtualHost for each external service
+	vhosts := make([]*VirtualHost, 0)
+	services := conf.Services.Services()
+	for _, service := range services {
+		if service.External {
+			if host := buildEgressHTTPRoute(service); host != nil {
+				vhosts = append(vhosts, host)
+			}
+		}
+	}
+
+	sort.Slice(vhosts, func(i, j int) bool { return vhosts[i].Name < vhosts[j].Name })
+
+	rConfig := &HTTPRouteConfig{VirtualHosts: vhosts}
+
+	listener := &Listener{
+		Address:    fmt.Sprintf("tcp://%s:80", WildcardAddress),
+		BindToPort: true,
+		Filters: []*NetworkFilter{
+			{
+				Type: "read",
+				Name: HTTPConnectionManager,
+				Config: HTTPFilterConfig{
+					CodecType:   "auto",
+					StatPrefix:  "http",
+					AccessLog:   []AccessLog{{Path: DefaultAccessLog}},
+					RouteConfig: rConfig,
+					Filters: []HTTPFilter{
+						{
+							Type:   "decoder",
+							Name:   "router",
+							Config: FilterRouterConfig{},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	listeners := []*Listener{listener}
+	clusters := rConfig.clusters().normalize()
+
+	clusters.setTimeout(conf.Mesh.ConnectTimeout)
+
+	return &Config{
+		Listeners: listeners,
+		Admin: Admin{
+			AccessLogPath: DefaultAccessLog,
+			Address:       fmt.Sprintf("tcp://%s:%d", WildcardAddress, conf.Mesh.ProxyAdminPort),
+		},
+		ClusterManager: ClusterManager{
+			Clusters: clusters,
+		},
+	}
+}
+
+// buildEgressRoute translates an egress rule to an Envoy route
+func buildEgressHTTPRoute(svc *model.Service) *VirtualHost {
+	var host *VirtualHost
+
+	// TODO Cluster names are not unique if external service definition contains more than one port
+	for _, servicePort := range svc.Ports {
+		protocol := servicePort.Protocol
+		switch protocol {
+		case model.ProtocolHTTP, model.ProtocolHTTP2, model.ProtocolGRPC:
+
+			route := &HTTPRoute{
+				Prefix:          "/",
+				Cluster:         buildEgressClusterName(svc.Address, servicePort.Port),
+				AutoHostRewrite: true,
```

Comment:
Should we just put `HostRewrite` to the external name? It's known ahead of time based on the cluster definition.